File: edgequake/sdks/python/examples/streaming_query.py
#!/usr/bin/env python3
"""
Streaming Query — EdgeQuake Python SDK

WHY: Streaming delivers tokens incrementally via Server-Sent Events (SSE),
enabling real-time UI updates without waiting for the full response.

Requirements:
    - EdgeQuake server running on http://localhost:8080
    - EDGEQUAKE_API_KEY environment variable set
    - Documents uploaded and indexed

Usage:
    export EDGEQUAKE_API_KEY="demo-key"
    python examples/streaming_query.py
"""
import os
import sys

from edgequake import EdgequakeClient


def main():
    client = EdgequakeClient(
        api_key=os.environ.get("EDGEQUAKE_API_KEY", "demo-key"),
        base_url=os.environ.get("EDGEQUAKE_URL", "http://localhost:8080"),
    )

    # ── 1. Streaming query via SSE ────────────────────────────

    # WHY: `client.query.stream()` returns an iterator that yields
    # parsed JSON chunks as they arrive from the server.
    print("Streaming query response:")
    try:
        for event in client.query.stream(
            query="Explain how knowledge graphs enhance RAG systems",
            mode="hybrid",
        ):
            # WHY: Each event is a parsed object from the SSE data line.
            # The exact shape depends on the backend's streaming format.
            if isinstance(event, dict) and "chunk" in event:
                sys.stdout.write(event["chunk"])
                sys.stdout.flush()
            elif isinstance(event, str):
                sys.stdout.write(event)
                sys.stdout.flush()
    except Exception as e:
        print(f"\n[Streaming error: {e}]")
    print("\n")

    # ── 2. Streaming chat (OpenAI-compatible) ─────────────────

    # WHY: Chat streaming follows the OpenAI delta format, making it
    # compatible with existing OpenAI-based UIs and libraries.
    print("Streaming chat response:")
    try:
        for chunk in client.chat.stream(
            model="edgequake",
            messages=[
                {
                    "role": "user",
                    "content": "What are the benefits of graph-based RAG?",
                },
            ],
        ):
            if isinstance(chunk, dict):
                delta = chunk.get("choices", [{}])[0].get("delta", {}).get("content")
                if delta:
                    sys.stdout.write(delta)
                    sys.stdout.flush()
    except Exception as e:
        print(f"\n[Streaming error: {e}]")
    print("\n")


if __name__ == "__main__":
    main()

File: edgequake/sdks/typescript/examples/streaming_query.ts
/**
 * Streaming Query — EdgeQuake TypeScript SDK
 *
 * WHY: Streaming delivers tokens incrementally via Server-Sent Events (SSE),
 * enabling real-time UI updates without waiting for the full response.
 *
 * Usage:
 *   npx tsx examples/streaming_query.ts
 */
import { EdgeQuake } from "@edgequake/sdk";

async function main() {
  const client = new EdgeQuake({
    baseUrl: process.env.EDGEQUAKE_URL ?? "http://localhost:8080",
    apiKey: process.env.EDGEQUAKE_API_KEY ?? "demo-key",
  });

  // ── 1. Streaming query via SSE ────────────────────────────

  // WHY: `client.query.stream()` returns an AsyncIterable that yields
  // parsed JSON chunks as they arrive from the server.
  console.log("Streaming query response:");
  for await (const event of client.query.stream({
    query: "Explain how knowledge graphs enhance RAG systems",
    mode: "hybrid",
  })) {
    // WHY: Each event is a parsed object from the SSE data line.
    // The exact shape depends on the backend's streaming format.
    if (event.chunk) {
      process.stdout.write(event.chunk);
    }
  }
  console.log("\n");

  // ── 2. Streaming chat (OpenAI-compatible) ─────────────────

  // WHY: Chat streaming follows the OpenAI delta format, making it
  // compatible with existing OpenAI-based UIs and libraries.
  console.log("Streaming chat response:");
  for await (const chunk of client.chat.stream({
    model: "edgequake",
    messages: [
      { role: "user", content: "What are the benefits of graph-based RAG?" },
    ],
  })) {
    const delta = chunk.choices?.[0]?.delta?.content;
    if (delta) {
      process.stdout.write(delta);
    }
  }
  console.log("\n");

  // ── 3. Streaming with abort ───────────────────────────────

  // WHY: AbortController lets you cancel streaming mid-flight,
  // useful for "stop generating" buttons in UIs.
  const controller = new AbortController();

  // Cancel after 3 seconds
  setTimeout(() => {
    console.log("\n[Aborting stream after 3 seconds]");
    controller.abort();
  }, 3000);

  console.log("Streaming with abort:");
  try {
    for await (const event of client.query.stream({
      query: "Write a detailed essay about knowledge graphs",
      mode: "hybrid",
      signal: controller.signal,
    })) {
      if (event.chunk) process.stdout.write(event.chunk);
    }
  } catch (err) {
    if (err instanceof DOMException && err.name === "AbortError") {
      console.log("\n[Stream aborted successfully]");
    } else {
      throw err;
    }
  }
}

main().catch(console.error);
